lib/navigation/control/navigator.cpp

```diff
@@ -127,4 +132,65 @@ core::Result Navigator::accelerometerUpdate(
   return core::Result::kSuccess;
 }
 
-}  // namespace hyped::navigation
+void Navigator::publishCurrentTrajectory()
+{
+  const auto topic     = core::MqttTopic::kTest;
+  auto message_payload = std::make_shared<rapidjson::Document>();
+  message_payload->SetObject();
+
+  auto trajectory = currentTrajectory();
+  if (!trajectory) { logger_.log(core::LogLevel::kFatal, "Failed to get current trajectory"); }
+
+  rapidjson::Value displacement(trajectory->displacement);
+  rapidjson::Value velocity(trajectory->velocity);
+  message_payload->AddMember("displacement", displacement, message_payload->GetAllocator());
+  message_payload->AddMember("velocity", velocity, message_payload->GetAllocator());
+
+  const core::MqttMessage::Header header{.timestamp = 0,
+                                         .priority  = core::MqttMessagePriority::kNormal};
+  const core::MqttMessage message{topic, header, message_payload};
+  mqtt_->publish(message, core::MqttMessageQos::kExactlyOnce);
+}
+
+void Navigator::run()
+{
+  mqtt_->subscribe(core::MqttTopic::kKeyence);
+  mqtt_->subscribe(core::MqttTopic::kOpticalFlow);
+  mqtt_->subscribe(core::MqttTopic::kAccelerometer);
+
+  while (true) {
+    mqtt_->consume();
+
+    auto keyence_msg       = mqtt_->getMessage();
+    auto optical_msg       = mqtt_->getMessage();
+    auto accelerometer_msg = mqtt_->getMessage();
```

Comment:
this won't work, each time you call `getMessage()` you get whatever the next message is. you then need to look at what topic that message is from, then do something with it. I think you'll want to store the value until you have a fresh acceleration and a fresh optical velocity measurement then run the filter over it once.

Easiest way I can think of is have `Navigator` have a `std::optional<std::uint16_t> most_recent_keyence_` etc, then you set each to a value when the message comes in, and if at least `most_recent_accerometer_` and `most_recent_optical_flow_` have values then run the filter and set them back to `std::nullopt` afterwards